### PR TITLE
feat(curriculum): add react state and hooks quiz

### LINKED ISSUE #56225: React (state, hooks, memoisation)

### PR DESCRIPTION
**Pull request by mitchellabrouwer**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #56225 

Thank you for considering this pull request. I look forward to your feedback.

### CODE REVIEW

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 1]
```diff
@@ -17,439 +17,469 @@ Answer all of the questions below correctly to pass the quiz.
...
 
 #### --text--
 
-Placeholder question
+What are React custom hooks typically used for?
 
 #### --distractors--
 
-Placeholder distractor 1
+Executing an apps most complex functionality.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:318]:
```suggestion
Executing an app's most complex functionality.
```

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 2]
```diff
@@ -17,439 +17,469 @@ Answer all of the questions below correctly to pass the quiz.
...
 
 #### --text--
 
-Placeholder question
+Which is NOT use for fetching data in React?
 
 #### --distractors--
 
-Placeholder distractor 1
+SWR
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:384]: Let's expand this into the full name.

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 3]
```diff
@@ -17,439 +17,469 @@ Answer all of the questions below correctly to pass the quiz.
...
 
 ---
 
-Placeholder distractor 3
+React Query
 
 #### --answer--
 
-Placeholder answer
+MUI
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:396]: This one too.

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 4]
```diff
@@ -17,439 +17,469 @@ Answer all of the questions below correctly to pass the quiz.
...
 
-Placeholder answer
+Functional component
 
 ### --question--
 
 #### --text--
 
-Placeholder question
+If I have a complex function inside a component, which hook can I use to memoize its result?
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:424]:
```suggestion
If you have a complex function inside a component, which hook can you use to memoize its result?
```

Second-person preferred.

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 5]
```diff
@@ -17,439 +17,469 @@ Answer all of the questions below correctly to pass the quiz.
...
 
-Placeholder answer
+`useMemo`
 
 ### --question--
 
 #### --text--
 
-Placeholder question
+The `useCallback` hook...
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:446]:
```suggestion
What is the `useCallback` hook?
```

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 6]
```diff
@@ -17,439 +17,469 @@ Answer all of the questions below correctly to pass the quiz.
...
 
 #### --distractors--
 
-Placeholder distractor 1
+Allows you to define a callback function.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:450]:
```suggestion
A hook that allows you to define a callback function.
```

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 7]
```diff
@@ -17,439 +17,469 @@ Answer all of the questions below correctly to pass the quiz.
...
 
 ---
 
-Placeholder distractor 2
+Is a custom hook not managed by React.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:454]:
```suggestion
A custom hook not managed by React.
```

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 8]
```diff
@@ -17,439 +17,469 @@ Answer all of the questions below correctly to pass the quiz.
...
 
 ---
 
-Placeholder distractor 3
+Checks an app's performance in development only.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:458]:
```suggestion
A hook that checks an app's performance in development only.
```

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 9]
```diff
@@ -17,439 +17,469 @@ Answer all of the questions below correctly to pass the quiz.
...
 
 #### --answer--
 
-Placeholder answer
+Ensures a function isn't re-created each time a component re-renders.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:462]:
```suggestion
A hook that ensures a function isn't re-created each time a component re-renders.
```

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

**Comment by mitchellabrouwer** [comment]
Thanks @naomi-lgbt, I've made the required updates.

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 10]
```diff
@@ -17,439 +17,469 @@ Answer all of the questions below correctly to pass the quiz.
...
 
 #### --distractors--
 
-Placeholder distractor 1
+No particular reason at all.
 
 ---
 
-Placeholder distractor 2
+To abstracts repetitive logic.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:146]:
```suggestion
To abstract repetitive logic.
```

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

**Comment by mitchellabrouwer** [comment]
Thanks @naomi-lgbt, I've committed the update.

**Review by naomi-lgbt — APPROVED** [review]

**Review by huyenltnguyen — COMMENTED** [review]
I have some concerns regarding the following questions:
- "Which of the following is NOT a global React state library?"
- "Which of the following lists only includes built in React hooks?"
- "Which is NOT use for fetching data in React?"

There is nothing wrong with the questions themselves, but I'm not sure if we want them in our quiz, considering the target audience. 

The questions are probably more suitable for experienced developers who are already familiar with React's built-in features and React ecosystem.

I think it's better to have the questions cover React concepts and test how well campers understand them, rather than asking campers to identify libraries.

**Comment by mitchellabrouwer** [comment]
Thanks for reviewing. I've taken your concerns into account and created replacement questions.

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 11]
```diff
@@ -17,439 +17,474 @@ Answer all of the questions below correctly to pass the quiz.
...
+const [certificates, setCertificates] = useState([]);
+```
+
+Which is the most reliable way to update array state using the `useState` hook?
 
 #### --distractors--
 
-Placeholder distractor 1
+```jsx
+setCertificates(previousItems => ...previousItems, 'Front End');
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:171]:
```suggestion
setCertificates(previousItems => ...previousItems, "Front End");
```

let's use double quotes for JSX to be uniform with the rest of the curriculum

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 12]
```diff
@@ -17,439 +17,474 @@ Answer all of the questions below correctly to pass the quiz.
...
+```
 
 ---
 
-Placeholder distractor 2
+```jsx
+setCertificates('Front End');
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:177]:
```suggestion
setCertificates("Front End");
```

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 13]
```diff
@@ -17,439 +17,474 @@ Answer all of the questions below correctly to pass the quiz.
...
+```
 
 ---
 
-Placeholder distractor 3
+```jsx
+setCertificates().append('Front End');
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:183]:
```suggestion
setCertificates().append("Front End");
```

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 14]
```diff
@@ -17,439 +17,474 @@ Answer all of the questions below correctly to pass the quiz.
...
+```
 
 #### --answer--
 
-Placeholder answer
+```jsx
+setCertificates(previousItems => [...previousItems, 'Front End']);
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:189]:
```suggestion
setCertificates(previousItems => [...previousItems, "Front End"]);
```

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 15]
```diff
@@ -17,439 +17,474 @@ Answer all of the questions below correctly to pass the quiz.
...
 
 #### --text--
 
-Placeholder question
+What is wrong with this function?
+
+```jsx
+function updateSpaceship() {
+  setSpaceship(previousState => ({
+    name: 'Discovery'
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:201]:
```suggestion
    name: "Discovery"
```

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 16]
```diff
@@ -17,439 +17,474 @@ Answer all of the questions below correctly to pass the quiz.
...
 ### --question--
 
 #### --text--
 
-Placeholder question
+How many times will this code print to the console?
+
+```jsx
+useEffect(() => {
+  console.log('Nice work!!');
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:296]:
```suggestion
  console.log(| Nice work!!");
```

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 17]
```diff
@@ -17,439 +17,474 @@ Answer all of the questions below correctly to pass the quiz.
...
 ---
 
-Placeholder distractor 3
+Single-page applications don't actually use a single page for routing.
 
 #### --answer--
 
-Placeholder answer
-
+The `react-router-dom` library is a popular choice to handle routing in React apps.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md:490]:
```suggestion
The `react-router-dom` library is a popular choice to handle routing in React apps.

```
let's not remove EOF empty lines

**Review by huyenltnguyen — APPROVED** [review]
LGTM 👍

@mitchellabrouwer Thank you for the PR and congratulations on your first contribution to this repo 🎉
We look forward to your next.

Happy contributing!

File: curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md [hunk 18]
```diff
@@ -293,7 +293,7 @@ How many times will this code print to the console?
 
 ```jsx
 useEffect(() => {
-  console.log('Nice work!!');
+  console.log(| Nice work!!");
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/quiz-react-state-and-hooks/66f1a417757b6ca4eecd89d6.md]:
```suggestion
  console.log("Nice work!!");
```

**Pull request merged**